### ISSUE TITLE
Psalm don't detect inherit property type

### ISSUE THREAD

**Issue by vjik** [vimeo/psalm#11319]
https://psalm.dev/r/b667f563eb

**Comment by danog** [comment]
This is correct, overridden properties must not be looser than the parent (it's an extension of the equivalent rule present in PHP, applied to docblock types as well).

**Comment by vjik** [comment]
> This is correct, overridden properties must not be looser than the parent (it's an extension of the equivalent rule present in PHP, applied to docblock types as well).

But for method parameters type inherits correctly: https://psalm.dev/r/5afbcd2ec8

**Comment by vjik** [comment]
"overridden properties must not be looser than the parent" - i agree, but see https://psalm.dev/r/b667f563eb :

Property `A::$var` is `non-empty-string`.
Why Psalm think that `B::$var` is `string`? Shouldn't he understand that `B::$var` also is `non-empty-string`?

**Comment by vjik** [comment]
@danog WDYT?

**Comment by danog** [comment]
>Why does it understand this in the parameters of the method (https://psalm.dev/r/5afbcd2ec8), but not in the properties of the class?

It doesn't, because loosening property types in method parameters is simply allowed.

**Comment by vjik** [comment]
@danog I'll try to explain my pain again =)

**Case with method parameter**

```php
class A
{    
    /**
     * @param non-empty-string $var
     */
    public function test(string $var): void {
        echo $var;
    }
}

final class B extends A {
    public function test(string $var): void {
        /** @psalm-trace $var */
        echo $var;
    }
}
```

Psalm without extra annotations in class B knows, that `$var` is `non-empty-string`. Not `string`. It's fine.


**Case with class property**

```
class A
{
    /**
     * @var non-empty-string
     */
    public string $var = 'test';
}

final class B extends A {
    public string $var = 'hello';
}
```

Psalm don't understand that `B::$var` is `non-empty-string`. Why?

**Comment by danog** [comment]
Okay I see your point now, will fix

**Comment by vjik** [comment]
Any news?
